clustermesh: Add hidden flag --allow-unsafe-policy-skb-usage

joestringer opened the pull request:
This hidden flag can allow existing users who are using a high
clustermesh ID in ENI environments to continue using it at their own
risk. Recently we added a check to startup to detect this potentially
problematic configuration and warn users, however it's now no longer
possible to run Cilium in the risky configuration. This puts users in a
tough spot of whether to run an outdated release or manually patch out
this newly introduced guardrail.

In the long term we should be able to lift the restriction around
usage of clustermesh ID in the datapath, but for now we should try to
encourage users not to configure Cilium this way while still ultimately
leaving the decision up to them about how to manage their environments.

Related: #35349
Related: #36309

To use this feature, you need the `extraConfig.allow-unsafe-policy-skb-usage` flag to be set:

```
...
cluster:
  id: 166
extraConfig:
  allow-unsafe-policy-skb-usage: "true"
```

I tested this feature using this diff:

```
diff --git a/contrib/testing/kind-common.yaml b/contrib/testing/kind-common.yaml
index 1be34fdebdde..e866a213dfbb 100644
--- a/contrib/testing/kind-common.yaml
+++ b/contrib/testing/kind-common.yaml
@@ -18,3 +18,7 @@ readinessProbe:
   failureThreshold: 9999
 startupProbe:
   failureThreshold: 9999
+cluster:
+  id: 166
+extraConfig:
+  allow-unsafe-policy-skb-usage: "true"
diff --git a/install/kubernetes/cilium/templates/validate.yaml b/install/kubernetes/cilium/templates/validate.yaml
index 37da6cd651f7..d8ce7ed1b0be 100644
--- a/install/kubernetes/cilium/templates/validate.yaml
+++ b/install/kubernetes/cilium/templates/validate.yaml
@@ -151,7 +151,7 @@
 {{ if and
     (ne (index .Values.extraConfig "allow-unsafe-policy-skb-usage") "true")
     (or (and (ge (int .Values.cluster.id) 128) (le (int .Values.cluster.id) 255)) (and (ge (int .Values.cluster.id) 384) (le (int .Values.cluster.id) 511)))
-    (or .Values.eni.enabled .Values.alibabacloud.enabled (eq .Values.cni.chainingMode "aws-cni")) -}}
+    (or .Values.eni.enabled .Values.alibabacloud.enabled (eq .Values.cni.chainingMode "aws-cni") true) -}}
   {{ fail "Cilium is currently affected by a bug that causes traffic matched by network policies to be incorrectly dropped when running in either ENI mode (both AWS and AlibabaCloud) or AWS VPC CNI chaining mode, if the cluster ID is 128-255 (and 384-511 when maxConnectedClusters=511). Please refer to https://github.com/cilium/cilium/issues/21330 for additional details." }}
 {{- end }}
 
diff --git a/pkg/clustermesh/types/option.go b/pkg/clustermesh/types/option.go
index fe746e82b85f..77372776eb91 100644
--- a/pkg/clustermesh/types/option.go
+++ b/pkg/clustermesh/types/option.go
@@ -69,7 +69,7 @@ func (c ClusterInfo) ValidateStrict() error {
 // ValidateBuggyClusterID returns an error if a buggy cluster ID (i.e., with the
 // 7th bit set) is used in combination with ENI IPAM mode or AWS CNI chaining.
 func (c ClusterInfo) ValidateBuggyClusterID(ipamMode, chainingMode string) error {
-       if (c.ID&0x80) != 0 && (ipamMode == ipamOption.IPAMENI || ipamMode == ipamOption.IPAMAlibabaCloud || chainingMode == "aws-cni") {
+       if (c.ID&0x80) != 0 && (ipamMode == ipamOption.IPAMENI || ipamMode == ipamOption.IPAMAlibabaCloud || chainingMode == "aws-cni" || true) {
                return errors.New("Cilium is currently affected by a bug that causes traffic matched " +
                        "by network policies to be incorrectly dropped when running in either ENI mode (both " +
                        "AWS and AlibabaCloud) or AWS VPC CNI chaining mode, if 
the cluster ID is 128-255 (and " +
```

joestringer committed:
clustermesh: Add --allow-unsafe-policy-skb-usage

This hidden flag can allow existing users who are using a high
clustermesh ID in ENI environments to continue using it at their own
risk. Recently we added a check to startup to detect this potentially
problematic configuration and warn users, however it's now no longer
possible to run Cilium in the risky configuration. This puts users in a
tough spot of whether to run an outdated release or manually patch out
this newly introduced guardrail.

In the long term we should be able to lift the restriction around
usage of clustermesh ID in the datapath, but for now we should try to
encourage users not to configure Cilium this way while still ultimately
leaving the decision up to them about how to manage their environments.

Signed-off-by: Joe Stringer <joe@cilium.io>

giorio94 commented on pkg/clustermesh/cell.go:
```diff
@@ -19,6 +20,24 @@ import (
+}
+
+func (cfg ClustermeshConfig) Flags(flags *pflag.FlagSet) {
+	flags.Bool("allow-buggy-clustermesh", false,
```
Hmm, I'd suggest changing the flag name to `allow-buggy-identities`, `allow-buggy-netpols`, or something like that. My feeling with the current wording is that it implies that the problem is clustermesh-specific, but it actually affects the traffic in the local cluster if the given cluster IDs are configured, regardless of whether clustermesh is used or not.

joestringer replied:
Yeah the naming of this flag is not ideal, I'll need to think about it. In some sense the `buggy` aspect helps to scare users but I don't like associating that word with a feature. Maybe the more generic term is `--allow-unsafe-policy-skb-usage` or something. The thing I am not sure about though is that if the mitigation for the user is to modify the clustermesh IDs for their environment, then it seems like it should use clustermesh in the name. I appreciate though that it's only a problem with the combination of clustermesh and policy.

giorio94 replied:
> The thing I am not sure about though is that if the mitigation for the user is to modify the clustermesh IDs for their environment, then it seems like it should use clustermesh in the name. I appreciate though that it's only a problem with the combination of clustermesh and policy.

You can actually set a ClusterID different from zero even if you are not using clustermesh at all. I guess that my main point here is that having clustermesh in the term implies that it affects cross-cluster traffic only, while this issue is actually about the netpol enforcement inside the local cluster (and possibly cross-cluster, by extension).

giorio94 commented on pkg/clustermesh/cell.go:
```diff
@@ -19,6 +20,24 @@ import (
 	"github.com/cilium/cilium/pkg/option"
 )
 
+type ClustermeshConfig struct {
```
Nit. The struct handling the `cluster-id/cluster-name` flags is defined in `pkg/clustermesh/types/option.go`. I'd suggesting moving this one there as well (although I reckon it may need to be a separate struct given that the other is used by multiple components). The name seems to imply again that this is clustermesh-specific, which is not.

giorio94 commented on pkg/clustermesh/cell.go:
```diff
@@ -35,12 +54,18 @@ var Cell = cell.Module(
 
 	cell.Config(common.DefaultConfig),
 	cell.Config(wait.TimeoutConfigDefault),
+	cell.Config(defaultConfig),
```
The ClusterInfo config is registered [here](https://github.com/cilium/cilium/blob/pr%2Fjoe%2Fallow-broken-clustermesh/daemon/cmd/cells.go#L261), so I'd probably move this one there as well for consistency.

joestringer replied:
I ended up declaring a new `Quirks` module specifically to combine the new flag and the logic to detect/handle this scenario. That way the quirks module can be included in the Hive from the top-level as you suggest. It's not otherwise obvious to me when we should make config part of the module itself vs. at the top level; for instance the two lines directly above this are specifying default config for this module.

I did however end up modifying where you linked in order to only include this quirks logic in the cilium-agent and not in the other binaries that consume clustermesh.

EDIT: I tried the separate module approach for a bit, but I ended up fighting Hive/Cell dependencies a bit too much so in the end I gave up on that approach and merged it back into the main clustermesh cell.

giorio94 commented on pkg/clustermesh/cell.go:
```diff
@@ -35,12 +54,18 @@ var Cell = cell.Module(
 
-	cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager) error {
-		return info.ValidateBuggyClusterID(dcfg.IPAM, cnimgr.GetChainingMode())
+	cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager, cfg Configuration) error {
```
Nit. It would be better to depend on the `ClustermeshConfig` (however it'll be called), rather than the full `Configuration` object.

joestringer replied:
The `Configuration` object here is to pull a logger from somewhere. I couldn't find a better handle for a logger. If you have an alternative suggestion then I'd be open to feedback, otherwise I'll keep that parameter just for the logger.

giorio94 reviewed (changes requested):
> The idea for this PR is to create a hidden flag that advanced users can opt into if they know what they're doing and accept the risks. At least this means they're not locked into an older release of Cilium, potentially missing out on other fixes like security updates etc. Do you think this is a reasonable middle ground?

Sounds reasonable to me, with the caveat that connectivity is probably completely broken in this configuration as soon as network policies are used. I've left a couple of comments inline, mostly about the flag naming.

One extra note is that the same validation is also [performed via helm](https://github.com/cilium/cilium/blob/pr%2Fjoe%2Fallow-broken-clustermesh/install/kubernetes/cilium/templates/validate.yaml#L151-L155), and should be probably relaxed as well:

joestringer committed:
helm: Support extraConfig.allowUnsafePolicySKBUsage

Let the user configure the new quirks.allowUnsafePolicySKBUsage flag and
have that also bypass the helm-level check for the same issue. See
GH-21330 for more information about the underlying limitation.

Tested by force-enabling the IPAM options in the Helm template plus
applying this configuration YAML:

cluster:
  id: 166
extraConfig:
  allow-unsafe-policy-skb-usage: "true"

Signed-off-by: Joe Stringer <joe@cilium.io>

giorio94 commented on pkg/clustermesh/cell.go:
```diff
@@ -39,8 +39,14 @@ var Cell = cell.Module(
-	cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager) error {
-		return info.ValidateBuggyClusterID(dcfg.IPAM, cnimgr.GetChainingMode())
+	cell.Config(types.DefaultQuirks),
+	cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager, cfg Configuration, quirks types.QuirksConfig) error {
```
```suggestion
	cell.Invoke(func(logger logrus.FieldLogger, info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager, quirks types.QuirksConfig) error {
```

Nit. I'd suggest depending explicitly on the logger, rather than on the broader cfg object.

joestringer replied:
How do we make sure the Logger is the right one, for example including the right subsystem etc? Do we need to declare a specific clustermesh logger type and provide it from the same Cell?

giorio94 replied:
That's automatically taken care of by hive, which always provides a logger (either logrus or slog) scoped for the module that requested it. The logger part of the `Configuration` structure is also already retrieved that way, without any further customization.

joestringer replied:
It looks like Hive/Cell magically populates this. Incremental patch, will push soon:

```
diff --git a/pkg/clustermesh/cell.go b/pkg/clustermesh/cell.go
index 9cc5ba6591f2..d3f2c7cd2ffe 100644
--- a/pkg/clustermesh/cell.go
+++ b/pkg/clustermesh/cell.go
@@ -5,6 +5,7 @@ package clustermesh
 
 import (
        "github.com/cilium/hive/cell"
+       "github.com/sirupsen/logrus"
 
        "github.com/cilium/cilium/daemon/cmd/cni"
        "github.com/cilium/cilium/pkg/clustermesh/common"
@@ -40,10 +41,10 @@ var Cell = cell.Module(
        metrics.Metric(common.MetricsProvider(subsystem)),
 
        cell.Config(types.DefaultQuirks),
-       cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager, cfg Configuration, quirks types.QuirksConfig) error {
+       cell.Invoke(func(info types.ClusterInfo, dcfg *option.DaemonConfig, cnimgr cni.CNIConfigManager, log logrus.FieldLogger, quirks types.QuirksConfig) error {
                err := info.ValidateBuggyClusterID(dcfg.IPAM, cnimgr.GetChainingMode())
                if err != nil && quirks.AllowUnsafePolicySKBUsage {
-                       cfg.Logger.WithError(err).Error("Detected clustermesh ID configuration that may cause connection impact")
+                       log.WithError(err).Error("Detected clustermesh ID configuration that may cause connection impact")
                        return nil
                }
                return err
```

giorio94 reviewed (approved):
Thanks!
